schematic/schemas/data_model_nodes.py

```diff
@@ -0,0 +1,232 @@
+from schematic.utils.validate_rules_utils import validate_schema_rules
+from schematic.schemas.curie import uri2curie, curie2uri
+
+
+class DataModelNodes():
+    def __init__(self, attribute_relationships_dict):
+        self.namespaces = dict(rdf=Namespace("http://www.w3.org/1999/02/22-rdf-syntax-ns#"))
+        self.data_model_relationships = DataModelRelationships()
+        self.value_relationships = self.data_model_relationships.define_value_relationships()
+        self.edge_relationships_dictionary = self.data_model_relationships.define_edge_relationships()
+        self.properties = self.get_data_model_properties(attr_rel_dict=attribute_relationships_dict)
+
+    '''    
+    def node_present(self, G, node_name):
+        if node_name in G.nodes():
+            return True
+        else:
+            return False
+    '''
+    def gather_nodes(self, attr_info: tuple) -> list:
+        """Take in a tuple containing attriute name and relationship dictionary, and find all nodes defined in attribute information.
+        Args:
+            attr_info, tuple: (Display Name, Relationships Dictionary portion of attribute_relationships dictionary)
+        Returns:
+            nodes, list: nodes related to the given node (specified in attr_info).
+        Note:
+            Extracting nodes in this fashion ensures order is preserved.
+        """
+        # retrieve a list of relationship types that will produce nodes.
+        self.node_relationships =list(self.edge_relationships_dictionary.values())
+
+        # Extract attribute and relationship dictionary
+        attribute, relationship = attr_info
+        relationships = relationship['Relationships']
+
+        nodes = []
+        if attribute not in nodes:
+            nodes.append(attribute)
+        for rel in self.node_relationships:
+            if rel in relationships.keys():
+                nodes.extend([node.strip()
+                                for node in relationships[rel]])
+        return nodes
+
+    def gather_all_nodes(self, attr_rel_dict: dict):
+        """Gather all nodes in the data model, in order.
+        Args:
+            attr_rel_dict, dict: generated in data_model_parser
+                {Attribute Display Name: {
+                        Relationships: {
+                                    CSV Header: Value}}}
+        Returns:
+            all_nodes, list: List of all node display names in the data model preserving order entered.
+        Note:
+            Gathering nodes in this fashion ensures order is preserved.
+        """
+        all_nodes = []
+        for attr_info in attr_rel_dict.items():
+            nodes = self.gather_nodes(attr_info=attr_info)
+            all_nodes.extend(nodes)
+        # Remove any duplicates preserving order
+        all_nodes = list(dict.fromkeys(all_nodes).keys())
+        return all_nodes
+
+    def get_rel_node_dict_info(self, relationship: str) -> tuple:
```

Comment:
This method is kind of problematic in that it says it returns a tuple, but it only returns one if both nested if statements are true. Should be `Optional[tuple[str, dict]]`